### ISSUE TITLE
Tensors

### ISSUE THREAD

**Issue by martinapugliese** [martinapugliese/tales-science-data#211]
Tensors are generalisations of vectors and matrices to higher dimensions and furnish a representation which is very convenient to working with neural networks. In fact, TensorFlow, the Google Machine Learning library, especially devoted, but not only, to deep learning, is built on top of them. 

Tensors are such that:

* a 0D tensor (tensor of rank 0) is a scalar (a number)
* a 1D tensor (tensor of rank 1) is a vector
* a 2D tensor (...) is a matrix
* a 3D tensor is ... a 3D tensor
* ...

Each dimension of a tensor is called an *axis*. Note that you can create tensors of any rank with Numpy! Note that from the mathematical and physical points of view, this is an abuse of language as tensors aren't simply higher order matrices, but show some mathematical properties and transformation laws.

When passing a training set to a deep model, it is customary to place each sample in an array and build a tensor as this representation is very convenient. 

Suppose you're working on some images task for instance. Images are matrices (when grayscale), that is, tensors of rank 2. The typical way to pass a training set of images to a deep network is via the construction of a tensor of rank 3 where all images are placed in an array, effectively where the first axis denotes the sample. Same with colour images, which are tensors of rank 3, the third dimension being the colour channel; a set of those is a tensor of rank 4 where the first axis is the sample, the second and third are height and width and the fourth the colour.

Note that Tensorflow uses the convention here described, Theano puts the colour channel on the second axis, and then height and width instead. This structure is employed for each sort of dimensionality: you build a tensor where the first axes stores each sample. Videos go on 5D tensors as each frame is an image!

### Operations

Operations on tensors are the scaled up versions of operations on vectors and matrices. 

### Tensor broadcasting

Broadcasting is the procedure that makes it possible to compute operations over tensors of different rank, like for instance an addition between a vector and a matrix. What broadcasting does is "extending" the smallest of the tensors by replicating it across the missing axes so that it matches the shape of the other tensor. For example, if you want to sum vector $v = (1, 2, 2)$ to matrix $A = \begin{bmatrix}
    2  & 3 & 1 \\
    1  & 1 & 1
\end{bmatrix}
$, you replicate the vector over the missing axis to build a matrix, effectively then summing $V = \begin{bmatrix}
    1  & 2 & 2 \\
    1  & 2 & 2
\end{bmatrix}$ to $A$, which yields $\begin{bmatrix}
    3  & 5 & 3 \\
    2  & 3 & 3
\end{bmatrix}$ .
